### PR TITLE
Add Newton review guidelines

### LINKED ISSUE #3509: Formalize and improve review guidance for agents and humans
AI Agents are performing a lot of the review work on Newton. To improve review quality, we should more clearly define what to look out during reviews for both agents and humans. In particular, we should turn our (draft) documents

* [Newton API Specification](https://docs.google.com/document/d/16uOPnOW4WsrgPLHNO7z_Cz8hqN0qM7-XLm_X5XyCIso/edit?tab=t.0#heading=h.wx4lxqb1pibt)
* [API audit](https://docs.google.com/spreadsheets/d/199fvR9b1vajddI-WXYsXu6e5jLtfJCz4oWWW8b2rwdk/edit?gid=264377098#gid=264377098)
* [Release audit](https://github.com/newton-physics/newton/tree/main/.claude/skills/release-audit)

into a practical design/style guidance to catch issues earlier during PR review by humans, personal AI agents, and coderabbit. This should be prominently referenced from Agents.md. We need to check how much the current .claude/skills/newton-api-design needs to be extended for that.

Detailed scratch notes:
- CodeRabbit rules that flag unnecessarily large or complex PRs, unclear explanations, verbosity, and likely maintenance burden

### PR DESCRIPTION
**Pull request by jcarius-nv**
## Description

Add `REVIEW_GUIDELINES.rst` as suggested review guidance for Newton,
particularly for AI-assisted review. The guide separates substantial concerns
into three independent axes:

- **Fit:** whether Newton should own the outcome and its durable cost is
  proportionate;
- **Requirements:** whether the agreed behavior is implemented correctly and
  supported by suitable evidence; and
- **Standards:** whether the change follows the source-code guidelines and
  other repository obligations.

Add repository review instructions for agents, configure CodeRabbit to apply
both guides, publish the review guide through the documentation, and connect
the release audit to the canonical coding, review, deprecation, and changelog
policies.

Human reviewers may continue to use their own process and judgment. The guide
provides shared criteria rather than prescribing a mandatory human workflow.

This builds on the source-code guidelines introduced by #3890, which is now
merged.

Closes #3509.

Cross-agent repository skill discovery is deferred to #3921.

## Checklist

- [x] New or existing tests cover these changes (not applicable; documentation
      and review configuration only)
- [x] The documentation is up to date with these changes
- [x] For user-facing changes, a fragment has been added by following the
      [changelog fragment instructions](https://github.com/newton-physics/newton/blob/main/changelog/README.md)

## Test plan

- `git diff --check`
- `uvx --python 3.12 pre-commit run -a`
- Validate `.claude/skills/code-review-newton` with the skill validator
- Parse `.coderabbit.yml` with PyYAML

## Summary by CodeRabbit

- **Documentation**
  - Added comprehensive source-code and review guidelines covering public APIs, testing, documentation, and contribution standards.
  - Expanded contributor documentation with clearer links for workflows, governance, implementation, and pull-request reviews.
  - Added source-code and review-guideline pages to the user documentation navigation.
  - Clarified API export, experimental API, and example validation requirements.
- **Chores**
  - Improved release-audit guidance and review automation configuration for more consistent assessments.
  - Added a changelog entry documenting the new review guidance.

### CODE REVIEW

File: .claude/skills/release-audit/SKILL.md [hunk 1]
```diff
@@ -10,6 +10,12 @@ allowed-tools: Bash(git log *) Bash(git show *) Bash(git grep *) Bash(git tag *)
 
 Generates a markdown audit of a Newton release for keep/defer decisions (or, in retrospective mode, for skill calibration). Three modes, auto-detected in Phase 1:
 
+Read `CODING_GUIDELINES.rst` and `REVIEW_GUIDELINES.rst` from the
+repository root before auditing. They define the canonical coding, API, and
+review policies. This skill performs the final cross-release reconciliation;
+it does not replace pull-request review, and a prior review pass does not make
+any audit phase optional.
+
```
[the remark this answers is not in the document]
Comment (jcarius-nv) [reply .claude/skills/release-audit/SKILL.md:21]: Addressed in `827aebae`. Retrospective audits now explicitly use the current guides to calibrate current audit policy against tag-pinned evidence, and they must not claim historical noncompliance unless the requirement also existed at the target tag.

File: .claude/skills/release-audit/SKILL.md [hunk 2]
```diff
@@ -596,7 +620,7 @@ Never pass `--public`. Never file a destination the user did not choose.
 - Towncrier insertion marker: `<!-- towncrier release notes start -->`.
 - CHANGELOG subsection headers: `### Added`, `### Changed`, `### Deprecated`, `### Removed`, `### Fixed`.
 - Symbol extraction from entry text: backtick-quoted `newton.X`, `newton.X.Y`, `ClassName.method`, bare `ClassName` (capitalized identifier), bare `snake_case_name()`. The FIRST backtick-quoted symbol in the bullet is usually the primary subject.
-- Migration-guidance regex (Phase 5a 📐): case-insensitive search for `use\s+\x60`, `in favor of`, `renamed? to`, `replaced? by`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
+- Migration-guidance regex (Phase 5a ⬆️): case-insensitive search for `use\s+\x60`, `in favor of`, `renamed? to`, `replaced? by`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
```
[the remark this answers is not in the document]
Comment (jcarius-nv) [reply .claude/skills/release-audit/SKILL.md]: Addressed in `827aebae`. The migration-guidance patterns now match both immediate forms such as `Renamed to` / `Replaced by` and symbol-bearing forms such as `Rename X to Y` / `Replace X by Y`. I also verified all four forms against the documented regex.

File: .claude/skills/release-audit/SKILL.md [hunk 3]
```diff
@@ -596,7 +623,7 @@ Never pass `--public`. Never file a destination the user did not choose.
 - Towncrier insertion marker: `<!-- towncrier release notes start -->`.
 - CHANGELOG subsection headers: `### Added`, `### Changed`, `### Deprecated`, `### Removed`, `### Fixed`.
 - Symbol extraction from entry text: backtick-quoted `newton.X`, `newton.X.Y`, `ClassName.method`, bare `ClassName` (capitalized identifier), bare `snake_case_name()`. The FIRST backtick-quoted symbol in the bullet is usually the primary subject.
-- Migration-guidance regex (Phase 5a 📐): case-insensitive search for `use\s+\x60`, `in favor of`, `renamed? to`, `replaced? by`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
+- Migration-guidance regex (Phase 5a ⬆️): case-insensitive search for `use\s+\x60`, `in favor of`, `rename(?:d)?(?:\s+\x60[^\x60]+\x60)?\s+to`, `replace(?:d)?(?:\s+\x60[^\x60]+\x60)?\s+by`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
```
[the remark this answers is not in the document]
Comment (jcarius-nv) [reply .claude/skills/release-audit/SKILL.md]: Addressed in `d8df5fa8`. The rename and replacement branches now require a backtick-quoted target after `to` or `by`. I verified four valid forms and confirmed that incomplete or unrelated continuations no longer match.

File: REVIEW_GUIDELINES.rst [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+differences; asynchronous execution and graph capture; batching and world
+isolation; determinism and ordering; serialization and importers; and
+performance-sensitive allocation, synchronization, and data transfer.
+
+Require evidence in proportion to the claim and risk. Tests should demonstrate
+observable behavior and meaningful regression boundaries rather than merely
+repeat the implementation. Performance, scalability, determinism, numerical
+accuracy, and compatibility claims need representative measurements or
+primary-source evidence with enough context to interpret them. Missing
+evidence is a question unless its absence is itself the risk.
```
Comment (eric-heiden) [inline REVIEW_GUIDELINES.rst]: > Missing evidence is a question unless its absence is itself the risk.

I don't understand what this means, can we simplify this?

**Review by eric-heiden — COMMENTED** [review]
Thanks, I think the guidelines look good! This is what I got from comparing a review from Codex following these guidelines versus not using them on several PRs:

They improved:
- architectural and project-fit coverage;
- separation of project ownership, requirements, and standards;
- explicit treatment of evidence and maintenance cost;
- handling of already-raised findings.

They did not improve raw defect discovery in this trial: the control (Codex without these review guidelines) found more concrete behavioral edge cases.
This is not enough evidence to conclude that the guidelines reduce correctness findings overall, but the risk seems real.

I suggest separating discovery order from reporting order: require a blind, correctness-first pass covering boundary conditions, state transitions, numerical invariants, backends, and performance before evaluating Fit or reading existing reviews. The final review can still present Fit first. I would also clarify that missing tests or documentation should not substitute for identifying the underlying behavioral defect.

This is what Codex recommended we should incorporate into the guidelines:
> Resolve and review the exact merge-base..head diff and complete commit list.
Read CODING_GUIDELINES.rst, the PR description, and linked requirements.
Do not read existing review comments yet.
>
> First perform an adversarial correctness pass. Trace changed inputs, state,
counts, offsets, ownership, and outputs through every supported execution
mode. Check zero/one/many, capacity boundaries, heterogeneous inputs,
toggle/reset/reuse and partial failures, CPU/CUDA, autodiff, determinism,
graph capture, and numerical invariants. For performance changes, quantify
setup and steady-state complexity, allocations, synchronization, transfers,
and whether benchmarks exercise the changed path. Run a minimal adversarial
probe when feasible.
>
> Freeze candidate findings with location, triggering input, mechanism,
observable impact, evidence, and confidence.
>
> Then review Fit and Standards independently. Missing tests, documentation, or
measurements must not replace the search for an underlying behavioral defect.
>
> Present Fit first, but report every supported material behavioral,
correctness, compatibility, and performance finding even when Fit fails.
Separate demonstrated defects from questions and evidence requests.

File: REVIEW_GUIDELINES.rst (hunk 4 above)
Comment (jcarius-nv) [reply REVIEW_GUIDELINES.rst]: Agreed. I simplified this in `b1ebc225` to: “Missing evidence should prompt a question. Treat its absence as a finding when the change cannot be reviewed safely without it.”

For example, if the PR description does not include test output, ask what was run. If a change to collision semantics has no regression test capable of detecting the failure, the missing test is itself a finding.

File: .claude/skills/release-audit/SKILL.md [hunk 5]
```diff
@@ -596,7 +623,7 @@ Never pass `--public`. Never file a destination the user did not choose.
 - Towncrier insertion marker: `<!-- towncrier release notes start -->`.
 - CHANGELOG subsection headers: `### Added`, `### Changed`, `### Deprecated`, `### Removed`, `### Fixed`.
 - Symbol extraction from entry text: backtick-quoted `newton.X`, `newton.X.Y`, `ClassName.method`, bare `ClassName` (capitalized identifier), bare `snake_case_name()`. The FIRST backtick-quoted symbol in the bullet is usually the primary subject.
-- Migration-guidance regex (Phase 5a 📐): case-insensitive search for `use\s+\x60`, `in favor of`, `renamed? to`, `replaced? by`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
+- Migration-guidance regex (Phase 5a ⬆️): case-insensitive search for `use\s+\x60`, `in favor of`, `rename(?:d)?(?:\s+\x60[^\x60]+\x60)?\s+to\s+\x60[^\x60]+\x60`, `replace(?:d)?(?:\s+\x60[^\x60]+\x60)?\s+by\s+\x60[^\x60]+\x60`, `switch to`, `migrate to`, `prefer\s+\x60` (where `\x60` matches a backtick).
```
[the remark this answers is not in the document]
Comment (jcarius-nv) [reply .claude/skills/release-audit/SKILL.md]: Agreed. Addressed in `b1ebc225`: every migration-guidance alternative now requires a complete, non-empty backtick-delimited replacement target. I also added an explicit near-miss calibration example and validated seven accepted forms against seven incomplete or unrelated forms.

**Comment by jcarius-nv** [comment]
@eric-heiden Thank you—this is thoughtful, and I agree the trial exposes a real failure mode even though it is not enough to establish a general effect. Our previous instructions conflated discovery order with reporting order and could anchor the review on Fit or existing findings too early.

I addressed this in `8593f0d9` at two levels:

- `REVIEW_GUIDELINES.rst` now calls for a correctness-first Requirements pass before Fit evaluation or reading existing comments when practical. The final review still presents Fit first, but must retain every supported material defect. It also states that missing tests, documentation, or measurements cannot substitute for finding the underlying behavioral problem.
- The Newton review skill now enforces the detailed sequence: resolve the exact diff and commits; read requirements and primary sources; perform the adversarial correctness pass; freeze candidate findings; then read existing reviews for deduplication and assess Fit and Standards independently. Final output distinguishes demonstrated defects from questions and evidence requests.

This keeps the human guidance concise while giving agents a deterministic anti-anchoring workflow.

File: REVIEW_GUIDELINES.rst (hunk 4 above)
Comment (eric-heiden) [reply REVIEW_GUIDELINES.rst]: Thanks for clarifying!

**Review by eric-heiden — APPROVED** [review]
Thanks for addressing my concerns! In my trials the new guidelines performed much better now.
